File: basic/java/1005. 继续(3n+1)猜想 (25).java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Scanner;

public class Main {
	public static void main(String args[]) {
		Scanner in=new Scanner(System.in);
		int n=in.nextInt();
		int[]a=new int[n];
		boolean[]mark=new boolean[10001];
		for(int i=0;i<n;++i) {
			a[i]=in.nextInt();
			int num=a[i];
			while(num!=1) {
				if(num%2!=0)num=num*3+1;
				num/=2;
				if(mark[num]==true)break;
				mark[num]=true;
			}
		}
		in.close();
		LinkedList<Integer>list=new LinkedList<Integer>();
		for(int i=0;i<n;++i) 
			if(mark[a[i]]==false)list.add(a[i]);
		int[]res=new int[list.size()];
		for(int i=0;i<res.length;++i)res[i]=list.get(i);
		Arrays.sort(res);
		for (int i = res.length - 1; i > 0; i--) 
			System.out.print(res[i] + " ");
		System.out.print(res[0]);
	}
}

File: basic/1005. 继续(3n+1)猜想 (25).cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
const int maxn=101;
bool num[maxn*100];
int main(){
	int n;
	cin>>n;
	int *a=new int[n];
	for(int i=0;i<n;++i){
		cin>>a[i];
		int m=a[i];
		while(m!=1){
			if(m%2)m=m*3+1;
			m/=2;
			if(num[m])break;
			num[m]=true;
			//cout<<m<<endl;
		}
	}
	vector<int> res;
	for(int i=0;i<n;++i)
		if(!num[a[i]])res.push_back(a[i]);
	sort(res.begin(),res.end());
	for(int i=res.size()-1;i>=0;--i){
		if(i!=res.size()-1)cout<<" ";
		cout<<res[i];
	}
	return 0;
}
